### PR TITLE
Cache AttributeKey hashCode

### PR DESCRIPTION
**Pull request by trask**
This turns out to be very helpful when looking up attribute keys in an attributes map.

I thought about using AutoValue's `@Memoize`, but the implementation for that is lazy double-checked lock (i.e. a volatile read), which seems unnecessary overhead in this case.

### CODE REVIEW

File: api/all/src/main/java/io/opentelemetry/api/common/AttributeKeyImpl.java [hunk 1]
```diff
@@ -5,12 +5,26 @@
...
+  private AttributeKeyImpl(AttributeType type, String key) {
+    if (type == null) {
+      throw new NullPointerException("Null type");
+    }
+    this.type = type;
+    if (key == null) {
+      throw new NullPointerException("Null key");
+    }
+    this.key = key;
+    this.hashCode = buildHashCode();
```
Comment (trask) [inline api/all/src/main/java/io/opentelemetry/api/common/AttributeKeyImpl.java:26]: I don't love calling instance method from constructor, but the `EqualsVerifier.withCachedHashCode()` check seems to force that.

Comment (jkwatson) [reply api/all/src/main/java/io/opentelemetry/api/common/AttributeKeyImpl.java:26]: You can't make it a static method and pass in the fields?

Comment (jkwatson) [reply api/all/src/main/java/io/opentelemetry/api/common/AttributeKeyImpl.java:26]: oh, well, I think I'd rather skip that equalsverifier check in the test. What is that test doing under the hood anyway?

**Review by anuraaga — APPROVED** [review]
EqualsVerifier automatically generates some instances with reflection and makes sure equals and hashcode are implemented correctly. As we've implemented it by hand here I think it's worth keeping. The instance method `buildHashCode` isn't ideal but seems ok too.

Though I wonder what the difference is between `withCachedHashCode` and `withIgnoredFields("hashCode")`.

**Comment by trask** [comment]
> `withIgnoredFields("hashCode")`

```
EqualsVerifier found a problem in class io.opentelemetry.api.common.AttributeKeyImpl.
-> Significant fields: equals relies on type, but hashCode does not.
  one has hashCode 1
  one has hashCode 1
```

**Review by jkwatson — APPROVED** [review]

**Comment by jkwatson** [comment]
needs a rebase to pick up the fixed collector image

**Pull request merged**